src/ConversationList/ConversationListItem.vala

```diff
@@ -83,13 +95,67 @@ public class Mail.ConversationListItem : VirtualizingListBoxRow {
         grid.attach (topic, 1, 1, 2, 1);
         grid.attach (messages, 3, 1, 1, 1);
 
+        var archive_affordance = new ArchiveAffordance (Gtk.Align.END);
+        archive_affordance.get_style_context ().add_class ("left");
+
+        var trash_affordance = new TrashAffordance (Gtk.Align.START);
+        trash_affordance.get_style_context ().add_class ("right");
+
+        carousel = new Hdy.Carousel () {
+            allow_scroll_wheel = false
+        };
+        carousel.add (archive_affordance);
+        carousel.add (grid);
+        carousel.add (trash_affordance);
+        carousel.scroll_to (grid);
+
         get_style_context ().add_class ("conversation-list-item");
-        add (grid);
+        child = carousel;
 
         show_all ();
+
+        button_release_event.connect ((e) => {
```

Comment:
While we’re here can we use an event controller? https://valadoc.org/gtk+-3.0/Gtk.GestureSingle.html

Reply:
TBH I've never managed to get Controllers to work in GTK3 :(

When I do this in the `ConversationListItem`:

```vala
        var gesture = new Gtk.GestureMultiPress (this) {
            button = Gdk.BUTTON_SECONDARY,
            propagation_phase = BUBBLE
        };
``` 

I never get a signal. Any ideas why that is?

Reply:
In Gtk3 widgets don't contain references to controllers, so you need to save your `gesture` to a field. Just add `private Gtk.GestureMultiPress gesture;` in class.

Reply:
Oh that makes sense then, probably over read it somewhere :(
Thank you, it's fixed now!